### PR TITLE
Raise a TypeError if a constraint returns non-boolean

### PR DESCRIPTION
**Pull request by maxhully**
This will catch the bug that we noticed a couple weeks ago where a constraint (in particular `refuse_new_splits`) is configured improperly and ends up returning a non-boolean value that resolves as truthy, making it seem like the partition passes the constraint when really it isn't.

Adding this check showed that BasicChain has this problem (wrongly configured `refuse_new_splits`), so I removed that constraint from BasicChain.

This new check will cause errors anywhere that we still have the wrong usage of `refuse_new_splits` hanging around. I think this still includes all of the config stuff. **We should talk about if we should wait to merge this until after Your State 2.0.**

I also added some tests assuring that `MarkovChain` returns the initial state in the first step and that `pipe_to_table` records the initial values in the first row.

### CODE REVIEW

**Comment by maxhully** [comment]
(I have some old commits from that ChainMap concept in here, but it didn't work so I took them back out.)

**Comment by drdeford** [comment]
Nice! state_report and generate_ensemble are the your state 2.0 files and neither of them is currently using refuse_new_splits although we are calculating county split entropies so most (much of, some, a few) of the runs will probably have well-behaved county data that we could use.

File: rundmcmc/chain.py [hunk 1]
```diff
@@ -54,15 +54,20 @@ def __next__(self):
                 else:
                     continue
 
-            self.state.parent = None
+            # Please delete this comment
```
Comment (rwbogl) [inline rundmcmc/chain.py]: :2nd_place_medal:

Comment (maxhully) [reply rundmcmc/chain.py]: Lol! That was from my "editing on my phone in the GitHub editor" phase, when I could create new lines but not delete any

**Comment by maxhully** [comment]
Also adding an `accepted` boolean field to all states that the chain yields. This closes #103 .

**Pull request merged**